### DISCUSSION TITLE
Modules in the Niffler

### DISCUSSION THREAD

**Discussion by sujayeendra98** [Emory-HITI/Niffler#314] [Q&A] [answered]
Hi all,

I am new to Niffler and started with simple module png extraction .
Need some information on cold extraction module..need any installation for module to be used.

Thanks.

**Comment by pradeeban** [comment, marked as answer]
Hi,
Please go through the README at https://github.com/Emory-HITI/Niffler
and then the README in,
https://github.com/Emory-HITI/Niffler/blob/master/modules/cold-extraction/

That should help you get started.
Please get back to us if you have more questions.

**Comment by aleksanderLisiecki** [comment]
Hi,
I come here with similar problem - I try to examine how Niffler work and I'm trying to get on demand dicom data from DCM4CHEE PACS (served on docker containers) into Niffler (I think it's cold extraction module). 
I found an MedAI presentation about Niffler (https://www.youtube.com/watch?v=oiQAjfYc9Sc) and it clarified me how Niffler works in theory, but I can't get it to work.
I don't know how to configure and to use the Niffler to read some example data from DCM4CHEE PACS.

Is there any more info or maybe some examples how Niffler works than README?

**Reply by pradeeban** [reply]
Hi, 

Thanks for your interest and the link. Let me see if I can help you.

Do you have data in your DCM4CHEE PACS (served on docker containers)? i.e., is your DCM4CHEE configured as a PACS with DICOM data in it (a storescp running on that side)? Niffler Cold-Extraction module just queries and retrieves data from a PACS that already has DICOM data in it. You will also need to configure the PACS to accept connections from Niffler (as in any DICOM connection).

**Reply by aleksanderLisiecki** [reply]
Thank you for answer!

I have uploaded example data, it's visible from dcm4chee GUI. I can access that data through python pydicom package (QIDO-RS and WADO-RS services from Dicom standard) so I assume that storescp is configured by default.

I've configured system.json by giving "QueryAet: localhost:11112" value (default storescp port).
I've set csv file with value:
'''
PatientID
214355
'''
"214355" - is PatientID stored on PACS

After setting that I run an cold-extraction module form Niffler GUI and got an message "Processing Data . . . . This Process may take several minutes or Hours. You can check logs !!". 
In logs there is only info about starting On-demand Extraction and nothing happened.
I think the process should not take to much time because in the PACS there is only that one patient with one series and one .dcm image, is there something I did wrong? 

I also tried to give Storage Folder value like "a/b/c/" but after running nothing has changed and these directories were not created. How  Storage Folder should be configured?

**Comment by pradeeban** [comment]
Thanks for getting back. You won't have to wait for more than a couple of minutes to get the images. The logs are correct (they indicate the Niffler cold-extraction module started fine).

I see 2 issues here.

* You are using the front-end module, it seems (as you referred to "Niffler GUI"). The front-end module is largely experimental, and we advise against using it. Indirectly, it is going to call the cold-extraction module to run and you better just run the cold-extraction module directly.
 
I have updated the README in the dev and master branches of the front-end module to indicate this now.

Suggestion: Please use the [cold-extraction](https://github.com/Emory-HITI/Niffler/tree/master/modules/cold-extraction) module directly, by following its README.

* However, the above alone will likely not solve your challenge.

You are running a DCM4CHEE as a PACS. But I don't see that you have configured it to accept queries from Niffler.

You need to configure both sides (sender and receiver) to query and receive DICOM images. So, in this case, DCM4CHEE is your PACS/sender (SrcAet) and Niffler is your receiver (QueryAet).

Pls configure both Niffler and DCM4CHEE.

This is how Niffler's configurations look like - https://github.com/Emory-HITI/Niffler/blob/master/modules/cold-extraction/system.json

You see SrcAet, QueryAet, and DestAet.

Unless you have configured DCM4CHEE to accept queries from anywhere, it will not accept queries from Niffler without specifically configuring DCM4CHEE to accept queries from Niffler.

I have not used DCM4CHEE as the src AET (So I cannot help you much with that). Our source is our enterprise PACS (with actual clinical DICOM images in them). So, those were specifically configured to accept queries from Niffler (stating its AE Title, hostname, and port address) and not accept queries from anywhere openly.

Please let us know how things go or if you need any further clarification from the Niffler end.

**Reply by aleksanderLisiecki** [reply]
I go with your's advice and I realized that my docker image was wrong (I did it by myself). 
I'm struggling to fix this and realized that maybe you have a docker image with Niffler. If so, could you share it?

**Reply by pradeeban** [reply]
We don't have a Docker image for Niffler, although we have thought about that. We have a Docker just for png-extraction module of Niffler, which is not going to help you much - https://github.com/Emory-HITI/Niffler/blob/master/modules/png-extraction/Dockerfile

By starting with creating a Docker image (when it does not exist), you are going to battle way too many variables (DCM4CHEE, Niffler, and Docker complexities). My suggestion would be to get it to work on your computer/server without Docker and turn to Docker only after you confirm your deployment to be working.

**Comment by H9660** [comment]
Hello sir!

I would like to work on this project in GSoC 2023. It is the right place to discuss the project as in 2023?

**Reply by pradeeban** [reply]
Yes, @H9660.

Please go through the code repository and feel free to ask if you have specific questions.

**Comment by Pushpak-Gangle55** [comment]
Hi ,
I am new to Niffler and want to contribute in app-layer , please guide me that from where i have to start.
